### PR TITLE
feat(config): convert datasource env to hostRules

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds datasource env parsing into hostRules.

## Context:

Allows hostRules to be configured more conveniently using env.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
This will need documenting, but I'll do so only once confirming that everyone likes this approach.

File: lib/config/env.ts [hunk 1]
```diff
@@ -86,14 +88,70 @@ export function getConfig(env: NodeJS.ProcessEnv): GlobalConfig {
     });
   }
 
-  if (env.DOCKER_USERNAME && env.DOCKER_PASSWORD) {
-    config.hostRules.push({
-      hostType: datasourceDocker.id,
-      username: env.DOCKER_USERNAME,
-      password: env.DOCKER_PASSWORD,
-    });
+  const datasources = getDatasourceList();
```
Comment (viceice) [inline lib/config/env.ts:91]:
```suggestion
  const datasources = new Set(getDatasourceList());
```

perf

File: lib/config/env.ts [hunk 2]
```diff
@@ -86,14 +88,70 @@ export function getConfig(env: NodeJS.ProcessEnv): GlobalConfig {
...
+  const fields = ['token', 'username', 'password'];
+
+  const hostRules: HostRule[] = [];
+
+  for (const envName of Object.keys(env).sort()) {
+    const splitEnv = envName.toLowerCase().split('_').filter(Boolean);
+    const prefix = splitEnv.shift();
+    if (datasources.includes(prefix)) {
```
Comment (viceice) [inline lib/config/env.ts:99]:
```suggestion
    if (datasources.has(prefix)) {
```

perf

**Review by viceice — COMMENTED** [review]
like this, some performance suggestion 🙃

File: lib/config/env.ts [hunk 3]
```diff
@@ -86,14 +87,59 @@ export function getConfig(env: NodeJS.ProcessEnv): GlobalConfig {
...
+  const datasources = getDatasourceList();
+  const fields = ['token', 'username', 'password'];
+
+  const hostRules: HostRule[] = [];
+
+  for (const envName of Object.keys(env).sort()) {
+    // Double underscore __ is used in place of hyphen -
+    const splitEnv = envName.toLowerCase().replace('__', '-').split('_');
+    const hostType = splitEnv.shift();
+    if (datasources.includes(hostType)) {
```
Comment (rarkins) [inline lib/config/env.ts:99]:
```suggestion
    if (datasources.has(hostType)) {
```

**Comment by rarkins** [comment]
@Chumper do you think this "double underscore maps to hyphen" approach works well? I recall you found a similar problem in Bundler

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Needs documentation (once the code wizards have agreed on a approach).
I can't write the docs for this, but I'll be happy to review the docs that somebody else creates. 😉

**Comment by rarkins** [comment]
Yes, switching back to Draft until I add docs

**Comment by Chumper** [comment]
> @Chumper do you think this "double underscore maps to hyphen" approach works well? I recall you found a similar problem in Bundler

yeah, the problem that in bash the hyphen `-` is not a valid character in the name:
```
export TEST-123=abc
export: not valid in this context: TEST-123
```
It is the same with the period `.`, that's why bundler chose double underscore `__` as replacement for the period.
We would need a replacement for the hyphen `-` as well as that is not uncommon in a URL.

If you decide to use double underscore as a replacement for the hyphen that is fine by me.
I would recommend to stick to Bundler convention and add the hyphen as well.

So `MY__TEST` would be `MY.TEST`
and for hyphen we could do three underscores.
So MY___TEST` would be `MY-TEST`

In the end it depends on what characters from the URI definition are not valid in a bash env name and we need to support.
The period `.` and hyphen `-` are the basic ones we should support.

**Comment by rarkins** [comment]
The underscore is not a valid character in host names so I thought it would work fine to have underscore replace dot and double underscore replace hyphen.  If we were to add an extra underscore to both scenarios like you suggest then what do we gain? I don't think the bundler approach is so widely used that we have to follow it.. just learn from it (eg hyphens)

**Comment by Chumper** [comment]
Yeah, it was just an example. I am fine with single underscore for dot and double for hyphen.

**Comment by rarkins** [comment]
Waiting on #9815

**Comment by HonkingGoose** [comment]
Removing `status:blocked` as PR #9815 is now merged into `main`.

**Review by HonkingGoose — COMMENTED** [review]
@rarkins Don't we still need documentation for this? Or did you decide we don't need any documentation after all?

**Comment by rarkins** [comment]
@HonkingGoose you're right! I forgot why I'd left it in Draft in the first place

File: lib/config/env.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import is from '@sindresorhus/is';
 
 import { PLATFORM_TYPE_GITHUB } from '../constants/platforms';
-import * as datasourceDocker from '../datasource/docker';
+import { getDatasourceList } from '../datasource';
 import { logger } from '../logger';
+import { HostRule } from '../types';
```
Comment (viceice) [inline lib/config/env.ts:6]:
```suggestion
import type { HostRule } from '../types';
```

**Review by viceice — COMMENTED** [review]
just a small issue

File: lib/config/env.ts [hunk 5]
```diff
@@ -86,14 +87,52 @@ export function getConfig(env: NodeJS.ProcessEnv): GlobalConfig {
...
+          hostRules.push(newRule);
+        }
+      }
+    }
   }
 
+  config.hostRules = [...config.hostRules, ...hostRules];
+
   // These env vars are deprecated and deleted to make sure they're not used
   const unsupportedEnv = [
```
Comment (JamieMagee) [inline lib/config/env.ts:137]: Everything that is old is new again 😂 

If you really wanted to, you could write a compatibility layer for these 😉

**Review by JamieMagee — DISMISSED** [review]

File: docs/usage/private-modules.md [hunk 6]
```diff
@@ -211,3 +211,48 @@ For instructions on this, see the above section on encrypting secrets for the Wh
 - Replace the existing public key in the HTML with the public key you generated in the step prior
 - Use the resulting HTML encrypt page to encrypt secrets for your app before adding them to user/repository config
 - Configure the app to run with `privateKey` set to the private key you generated above
+
+### hostRules Configuration using Environment Variables
+
+For self-hosted users, it is possible to use environment variables to configure the most common type of `hostRules` for authentication.
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:217]:
```suggestion
Self-hosted users can use environment variables to configure the most common types of `hostRules` for authentication.
```

File: docs/usage/private-modules.md [hunk 7]
```diff
@@ -211,3 +211,48 @@ For instructions on this, see the above section on encrypting secrets for the Wh
...
+
+The format of the environment variables should be all upper-case and follow:
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:219]:
```suggestion
The format of the environment variables must be all upper-case and follow:
```

File: docs/usage/private-modules.md [hunk 8]
```diff
@@ -211,3 +211,48 @@ For instructions on this, see the above section on encrypting secrets for the Wh
...
+
+- Datasource name (e.g. `NPM`, `PYPI`)
+- Underscore (`_`)
+- `matchHost`
+- Underscore
+- Field name (`TOKEN`, `USER_NAME`, or `PASSWORD`)
+
+Hyphens (`-`) in datasource or host name should be replaced with double underscores (`__`) and periods (`.`) in host names replaced with single underscore (`_`).
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:227]:
```suggestion
Hyphens (`-`) in datasource or host name must be replaced with double underscores (`__`).
Periods (`.`) in host names must be replaced with a single underscore (`_`).
```

Might make more sense to split this over two sentences?

File: docs/usage/private-modules.md [hunk 9]
```diff
@@ -211,3 +211,48 @@ For instructions on this, see the above section on encrypting secrets for the Wh
...
+### hostRules Configuration using Environment Variables
+
+For self-hosted users, it is possible to use environment variables to configure the most common type of `hostRules` for authentication.
+
+The format of the environment variables should be all upper-case and follow:
+
+- Datasource name (e.g. `NPM`, `PYPI`)
+- Underscore (`_`)
+- `matchHost`
+- Underscore
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:224]:
```suggestion
- Underscore (`_`)
```

File: docs/usage/private-modules.md [hunk 10]
```diff
@@ -211,3 +211,48 @@ For instructions on this, see the above section on encrypting secrets for the Wh
 - Replace the existing public key in the HTML with the public key you generated in the step prior
 - Use the resulting HTML encrypt page to encrypt secrets for your app before adding them to user/repository config
 - Configure the app to run with `privateKey` set to the private key you generated above
+
+### hostRules Configuration using Environment Variables
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:215]:
```suggestion
### hostRules configuration using environment variables
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some small changes. 😄

File: docs/usage/private-modules.md [hunk 11]
```diff
@@ -211,3 +211,49 @@ For instructions on this, see the above section on encrypting secrets for the Wh
...
+- `matchHost`
+- Underscore (`_`)
+- Field name (`TOKEN`, `USER_NAME`, or `PASSWORD`)
+
+Hyphens (`-`) in datasource or host name must be replaced with double underscores (`__`).
+Periods (`.`) in host names must be replaced with a single underscore (`_`).
+
+Examples:
+
+`NPM_REGISTRY_NPJS_ORG_TOKEN=abc123`:
```
Comment (HonkingGoose) [inline docs/usage/private-modules.md:232]:
```suggestion
`NPM_REGISTRY_NPMJS_ORG_TOKEN=abc123`:
```

I think we're missing a letter `M` in the `matchHost` part here?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — DISMISSED** [review]

File: docs/usage/private-modules.md [hunk 12]
```diff
@@ -211,3 +211,49 @@ For instructions on this, see the above section on encrypting secrets for the Wh
...
+
+### hostRules configuration using environment variables
+
+Self-hosted users can use environment variables to configure the most common types of `hostRules` for authentication.
+
+The format of the environment variables must be all upper-case and follow:
+
+- Datasource name (e.g. `NPM`, `PYPI`)
+- Underscore (`_`)
+- `matchHost`
```
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/private-modules.md:223]: Yes, optional. But right now NPM_TOKEN won't be added to any `.npmrc` (I think we should add that independently)

**Review by viceice — APPROVED** [review]

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I feel we're missing some `h4` headings to separate out the different examples.

How about something like this?

```markdown
#### GitHub example

Renovate config JSON

#### GitLab example

Renovate config JSON

#### Docker example

Renovate config JSON
```

**Comment by rarkins** [comment]
> I feel we're missing some `h4` headings to separate out the different examples.
> 
> How about something like this?

Would be great if you can improve it in a separate PR, as I really want to get this one merged finally.

**Pull request merged**